test-lib.sh

```diff
@@ -311,6 +311,49 @@ EOF
   : "  Success!"
 }
 
+# ct_check_exec_env_vars
+# --------------------
+# Checks if all relevant environment variables from `docker run`
+# can be found in `docker exec` as well.
+# Uses: $CID_FILE_DIR - path to directory containing cid_files
+# Uses: $IMAGE_NAME - name of the image being tested
+function ct_check_exec_env_vars() {
+  local tmpdir exec_envs run_envs cid old_IFS
+  local var_name stripped filtered_envs
+  tmpdir=$(mktemp -d)
+  CID_FILE_DIR=${CID_FILE_DIR:-$(mktemp -d)}
+  # Get environment variables from `docker run`
+  docker run --rm "$IMAGE_NAME" /bin/bash -c "env" | sort > "$tmpdir/run_envs"
+  # Get environment variables from `docker exec`
+  ct_create_container "test_exec_envs" bash -c "sleep 1000" >/dev/null
+  cid=$(ct_get_cid "test_exec_envs")
+  docker exec "$cid" env | sort > "$tmpdir/exec_envs"
+  # Filter out variables we are not interested in
+  # Always check X_SCLS, ignore PWD
+  # Check variables from `docker run` that have alternative paths inside (/opt/rh, /opt/app-root)
+  exec_envs=$(cat "$tmpdir/exec_envs")
+  run_envs=$(grep "^X_SCLS=\|/opt/rh\|/opt/app-root" "$tmpdir/run_envs" | grep -v "^PWD=")
+  for variable in $run_envs; do
+    var_name=$(echo "$variable" | awk -F= '{ print $1 }')
+    stripped=$(echo "$variable" | awk -F= '{ print $2 }')
+    filtered_envs=$(echo "$exec_envs" | grep "^$var_name=")
+    [ -z "$filtered_envs" ] && { echo "$var_name not found during \` docker exec\`"; return 1; }
+    old_IFS=$IFS
+    # For each such variable compare its content with the `docker exec` result, use `:` as delimiter
+    IFS=:
+    for value in $stripped; do
+        if [ -n "${filtered_envs##*$value*}" ]; then
+            echo " Value $value is missing from variable $var_name"
+            return 1
+        fi
+    done
+  done
+  IFS=$old_IFS
+  echo " All values present in \`docker exec\`"
+  ct_cleanup
```

Comment:
I'm not sure whether it is a good idea to run `ct_cleanup` in one of the library functions like this, since it has effect on created containers that were not created in this function. It would remove them and displayed its log, but it would be confusing because the other containers are not related to this test. I'd suggest to rather either remove just the single created container or leave it up to the calling script to run `ct_cleanup`.

Reply:
Maybe we should make clear in the comment of `ct_cleanup` function how it is supposed to be used (whether it can be called multiple times). And we can also divide it to two functions: `ct_cleanup` working with all previously created containers, and `ct_cleanup_single` to cleanup only a single container.

Reply:
Thanks for catching this! I wanted to move the call to nodejs testuite anyway, so let us do that. Makes more sense to do cleanup with log dumps after the test suite completes anyway.